### PR TITLE
Implement multiway DoF identities on vertices/lines/quads.

### PR DESCRIPTION
**Pull request by bangerth**
This is the first step in addressing #12626. @marcfehling You will be interested in this. I think it's also an elegant algorithm :-)

/rebuild

### CODE REVIEW

File: tests/hp/hp_line_dof_identities_mixed_multiway.cc [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+  const auto identities = fe_collection.hp_line_dof_identities(fe_indices);
+
+  for (unsigned int i = 0; i < identities.size(); ++i)
+    {
+      deallog << "Identity set #" << i << std::endl;
+      for (const auto &p : identities[i])
+        {
+          deallog << "  " << fe_collection[p.first].get_name()
+                  << ": line dof index " << p.second << std::endl;
```
Comment (marcfehling) [inline tests/hp/hp_line_dof_identities_mixed_multiway.cc:59]:
```suggestion
      deallog << "Identity set #" << i;
      for (const auto &p : identities[i])
        {
          deallog << std::endl << "  " << fe_collection[p.first].get_name()
                  << ": line dof index " << p.second;
```
Here and in the following: You can remove the blank line in the output file by writing it like this, but this is really not important.

File: tests/hp/hp_line_dof_identities_mixed_multiway.cc [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+
+template <int dim>
+void
+test()
+{
+  deallog << "dim=" << dim << std::endl;
+
+  hp::FECollection<dim> fe_collection;
+  for (unsigned int i = 1; i < 8; ++i)
```
Comment (marcfehling) [inline tests/hp/hp_line_dof_identities_mixed_multiway.cc:40]:
```suggestion
  for (unsigned int i = 1; i <= 8; ++i)
```
Here and in the other tests: Could you add `FE_Q(8)` as well? This way, we would end up with more sets of identities just between the `FE_Q(4)` and `FE_Q(8)` elements (two more for lines, eight more for quads).

**Review by marcfehling — APPROVED** [review]
I like this generalization. Not only can we use it for the hanging node algorithm, but also for the unification algorithm. Maybe it makes sense to pre-compute these identities and cache them (for a later PR).

I just have requests for cosmetic changes and would like to include `degree=8` elements in the tests.

File: tests/hp/hp_line_dof_identities_mixed_multiway.cc [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    {
+      deallog << "Identity set #" << i << std::endl;
+      for (const auto &p : identities[i])
+        {
+          deallog << "  " << fe_collection[p.first].get_name()
+                  << ": line dof index " << p.second << std::endl;
+        }
+    }
+
+  deallog << std::endl;
```
Comment (marcfehling) [inline tests/hp/hp_line_dof_identities_mixed_multiway.cc:63]: Here and later: Do you want to keep this blank line? I find it rather irritating, but it doesn't really matter.
```suggestion
```

**Comment by bangerth** [comment]
Thanks, @marcfehling . I've made the requested modifications. It turns out that we don't compute the two additional identities between FE_Q(4) and FE_Q(8). It took me a while to figure out why this is so: We don't use equidistant support points for `FE_Q`, but the Gauss-Lobatto points and there are no equivalences other than the midpoint.

I will add another test as a follow-up that checks the identity for equidistant support points.

**Comment by bangerth** [comment]
I added the tests here, using equidistant elements. I think the additional identities are correct.

**Comment by marcfehling** [comment]
> We don't use equidistant support points for `FE_Q`, but the Gauss-Lobatto points and there are no equivalences other than the midpoint.

Interesting! I didn't know that the support points aren't equidistant.

Maybe I should adjust the support point location for the `pgfplots` tool we use in the paper as well -- using the result of [Polynomials::jacobi_polynomial_roots()](https://www.dealii.org/developer/doxygen/deal.II/namespacePolynomials.html#adc782b64b1a5c3d009eb24dd372be6e4), but I guess for demonstration purposes equidistant points are good enough :)

**Comment by marcfehling** [comment]
> I think the additional identities are correct.

Yes, to my understanding, they are correct.

**Review by marcfehling — APPROVED** [review]

**Comment by bangerth** [comment]
If it isn't too much work, playing with moving the points might actually be a nice touch.

**Pull request merged**